pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go

```diff
@@ -0,0 +1,425 @@
+
+	derivedEndpointSlice.Labels = maps.Clone(derivedService.Labels)
+	if derivedEndpointSlice.Labels == nil {
+		derivedEndpointSlice.Labels = map[string]string{}
+	}
+	derivedEndpointSlice.Labels[mcsapiv1alpha1.LabelServiceName] = localEndpointSlice.Labels[discoveryv1.LabelServiceName]
+	derivedEndpointSlice.Labels[discoveryv1.LabelServiceName] = derivedService.Name
+	derivedEndpointSlice.Labels[mcsapiv1alpha1.LabelSourceCluster] = clusterName
+	derivedEndpointSlice.Labels[discoveryv1.LabelManagedBy] = endpointSliceLocalMCSAPIControllerName
+
+	derivedEndpointSlice.Annotations = maps.Clone(derivedService.Annotations)
+	derivedEndpointSlice.AddressType = localEndpointSlice.AddressType
+
+	// We need to deep copy the slices as those struct are full of pointers
+	derivedEndpointSlice.Endpoints = make([]discoveryv1.Endpoint, 0, len(localEndpointSlice.Endpoints))
+	for _, endpoint := range localEndpointSlice.Endpoints {
+		derivedEndpointSlice.Endpoints = append(derivedEndpointSlice.Endpoints, *endpoint.DeepCopy())
+	}
+	derivedEndpointSlice.Ports = make([]discoveryv1.EndpointPort, 0, len(localEndpointSlice.Ports))
+	for _, port := range localEndpointSlice.Ports {
+		derivedEndpointSlice.Ports = append(derivedEndpointSlice.Ports, *port.DeepCopy())
+	}
+}
+
+func newDerivedEndpointSlice(
+	localEndpointSlice *discoveryv1.EndpointSlice,
+	clusterName string, derivedService *corev1.Service,
+) *discoveryv1.EndpointSlice {
+	derivedEndpointSlice := discoveryv1.EndpointSlice{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:         derivedService.Name + "-" + getSuffix(localEndpointSlice),
+			GenerateName: derivedService.Name + "-",
+			Namespace:    derivedService.Namespace,
+		},
+	}
+
+	updateDerivedEndpointSlice(&derivedEndpointSlice, localEndpointSlice, clusterName, derivedService)
+	return &derivedEndpointSlice
+}
+
+func (r *mcsAPIEndpointSliceMirrorReconciler) addEndpointSlices(
+	ctx context.Context, toAddSuffixes []string, localEndpointSlicesMap map[string]*discoveryv1.EndpointSlice,
+	derivedService *corev1.Service,
+) error {
+	if len(toAddSuffixes) == 0 {
+		return nil
+	}
+
+	for _, suffix := range toAddSuffixes {
+		localEndpointSlice := localEndpointSlicesMap[suffix]
+
+		if err := r.Client.Create(ctx, newDerivedEndpointSlice(localEndpointSlice, r.clusterName, derivedService)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *mcsAPIEndpointSliceMirrorReconciler) deleteEndpointSlices(
+	ctx context.Context, toDeleteSuffixes []string, derivedEndpointSlicesMap map[string]*discoveryv1.EndpointSlice,
+) error {
+	if len(toDeleteSuffixes) == 0 {
+		return nil
+	}
+
+	for _, suffix := range toDeleteSuffixes {
+		if err := r.Client.Delete(ctx, derivedEndpointSlicesMap[suffix]); client.IgnoreNotFound(err) != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+const (
+	logAttrReq              = "req"
+	logAttrToUpdateSuffixes = "toUpdateSuffixes"
+	logAttrToAddSuffixes    = "toAddSuffixes"
+	logAttrToDeleteSuffixes = "toDeleteSuffixes"
+)
```

Comment:
Nit. Most of the log attributes are currently defined in `pkg/logging/logfields.go`. It could make sense moving these there as well, to favor consistency and reuse.

Reply:
The "req" is already there that's true, for the other three should I add it there even though it's kind off specific to this controller?

Reply:
Both approaches should work, I agree they are difficult to reuse.